learning.py

```diff
@@ -810,7 +815,110 @@ def init_examples(examples, idx_i, idx_t, o_units):
 
 
 def find_max_node(nodes):
-    return nodes.index(argmax(nodes, key=lambda node: node.value))
+    return nodes.index(max(nodes, key=lambda node: node.value))
+
+
+class BinarySVM:
+    def __init__(self, kernel=linear_kernel, C=1.0):
+        self.kernel = kernel
+        self.C = C
+        self.eps = 1e-6
+        self.n_sv = -1
+        self.sv_x, self.sv_y, = np.zeros(0), np.zeros(0)
+        self.alphas = np.zeros(0)
+        self.w = None
+        self.b = 0.0  # intercept
+
+    def fit(self, X, y):
+        # In QP formulation (dual): m variables, 2m+1 constraints (1 equation, 2m inequations)
+        self.QP(X, y)
+        sv_indices = list(filter(lambda i: self.alphas[i] > self.eps, range(len(y))))
+        self.sv_x, self.sv_y, self.alphas = X[sv_indices], y[sv_indices], self.alphas[sv_indices]
+        self.n_sv = len(sv_indices)
+        if self.kernel == linear_kernel:
+            self.w = np.dot(self.alphas * self.sv_y, self.sv_x)
+        # calculate b: average over all support vectors
+        sv_boundary = self.alphas < self.C - self.eps
+        self.b = np.mean(self.sv_y[sv_boundary] - np.dot(self.alphas * self.sv_y,
+                                                         self.kernel(self.sv_x, self.sv_x[sv_boundary])))
+
+    def QP(self, X, y):
```

Comment:
In this function, isn't it possible to rename the variables to something more descriptive? I am not familiar with SVM literature, but I think it would be best if the variable names described a bit more what each variable does.

Reply:
The name of the variables is common in the SVM scope, anyway I added doc.